toolz/itertoolz.py

```diff
@@ -855,3 +855,25 @@ def diff(*seqs, **kwargs):
             vals = tuple(map(key, items))
             if vals.count(vals[0]) != N:
                 yield items
+
+
+def topk(k, seq, key=None):
+    """
+    Find the k largest elements of a sequence
+
+    Operates lazily in ``n*log(k)`` time
+
+    >>> topk(2, [1, 100, 10, 1000])
+    (1000, 100)
+
+    Use a key function to change sorted order
+
+    >>> topk(2, ['Alice', 'Bob', 'Charlie', 'Dan'], key=len)
+    ('Charlie', 'Alice')
+    """
+    if key is None:
+        return tuple(heapq.nlargest(k, seq))
+    if not callable(key):
+        key = getter(key)
+    keyed_seq = map(lambda x: (key(x), x), seq)
```

Comment:
When a key function is passed, is it expected that items with equal key values will be in the same order as the original sequence?  This is not currently the case.  We would need to change `(key(x), x)` to something like `(key(x), i, x)` where `i` is incrementing.